Add mixed precision and 16-bit support to `TorchForecastingModel`

daidahao opened the pull request:
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

<!-- Please mention an issue this pull request addresses. -->
Addresses #2484 #2755

### Summary

<!-- Provide a general description of the code changes in your pull
request. If your pull request is not ready to merge, please create
a draft and ask for comments. -->

Previously, the `"precision"` option set in `pl_trainer_kwargs` is ignored in favour of precision of the time series data type. That led to the lack of support for mixed precision training and 16-bit true precision despite `pytorch-lightning.Trainer` native support.

I added support for mixed precision and any other precision options supported by Lightning in `TorchForecastingModel`.

Here, by default when the `"precision"` option is not set in `pl_trainer_kwargs` or a custom trainer, the model precision is determined by the data type. When it is set, we can assume that the user understand the ramifications of that option and the option is always passed along to the trainer, even when there is a mismatch between the model precision and the data type.

Sometimes the mismatch is necessary when the data type is 32-bit but the model can be trained in mixed precision. The other times the mismatch would cause an error, say, training 32-bit model on 64-bit data.

**Precision is an advanced option for `TorchForecastingModel` training and users should only use it at their own benefits and risks!**


### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, code samples, or others. -->

<!--Thank you for contributing to darts! -->
N/A

dennisbader commented:
Ahh got it, thanks @daidahao. I'll push the update in a few minutes.

dennisbader reviewed (changes requested):
Thanks for this PR @daidahao, it's looking like a great start 🚀 

I tested it out a little and I do have some concerns / findings:

- I'm testing on a Apple M1 Pro, 2021, macOS 15.3.2 (tested CPU and MPS (GPU)).
  - 16-true: forecasts contain **only NaN values** regardless of input series dtype or CPU / GPU. Not exactly sure exactly if it's a hardware issue or code. How does it look on your side?
  - bf16-true: **works** but sadly we have to convert to np.float32 in the end 😢 
  - bfloat16-mixed and float16-mixed: 
    - On MPS (GPU):
      - When input has dtype np.float16: **Kernel dies**
      - When input has dtype np.float32: **Works** (but converts output to np.float32).
     - On CPU: 
        - When input has dtype np.float16: **Raises error** `RuntimeError: Unexpected floating ScalarType in at::autocast::prioritize` (since `torch.cat()` doesn't support float16 on CPU).
        - When input has dtype np.float32: **Works** (but converts output to np.float32).

Looking at the above, the only configurations that seems to work are:

- bf16-true regardless of input `series` data type (float16, float32, float64).
- bf16-mixed and 16-mixed when input `series` data type is float32.

Given all of the above it seems like there are a lot of exceptions happening which can be quite tricky for the user to understand. So my ideas:

- Should we for now only handle these and otherwise raise exceptions from our side? This could help users a lot (even though you are correct that it is an advanced feature)

daidahao commented:
Hi Dennis,

I've tested it again on both MacBook Pro (M4 Pro 48G) and a GPU server (Tesla V100 16G). I have dependencies close to yours.

Here are my testing results with TSMixerModel:
- 16-true: NaN losses and predictions regardless of the input series type.
- 16-mixed:
   - MPS: disabled by Lightning and using 32-true due to not using CUDA
   - CPU: switched to bf16-mixed instead by Lightning
   - GPU (cuda): works with both 16-bit and 32-bit input.
- bf16-true: works with both 16-bit and 32-bit input.
- bf16-mixed: 
  - MPS: not supported for either 16-bit or 32-bit input.
  - CPU: same as yours, error for 16-bit input and works with 32-bit input
  - GPU (cuda): works with both 16-bit and 32-bit input.
 
Lightning documentation mentioned that BFloat16 is more stable than 16-true and the results here seem to confirm that. Given the results, can I suggest defaulting to bf16-mixed and producing a warning in `_init_model()` when input is 16-bit and precision not set by the user?

Conversion from bfloat16 to np.float32 is probably fine since numpy does not have an equivalent of bfloat16. We could produce a warning to user when `predict_from_dataset()` is called on 16-bit input so they know that the data type has changed from the input.

To you last point, if we handle some precision options for the user, the support from Lightning might change in the future and we could end up with more nested conditionals to deal with possible combinations of device, input type, and precision.

Finally, I've noticed that some models do not support any mixed precision like TFT due to masking operations.

Here is my script for testing the precision option.

```python
import numpy as np
import pandas as pd
from darts import TimeSeries
from darts.models import TSMixerModel

tfm_kwargs = {
    "batch_size": 16,
    "n_epochs": 1,
    "random_state": 42,
    "pl_trainer_kwargs": {
        "accelerator": "mps",
        "precision": "16-mixed",
        "enable_progress_bar": False,
        "enable_model_summary": False,
    },
}
length = 10000

times = pd.date_range("20130101", periods=length, freq="s")
pd_series = pd.Series(np.linspace(0, 1, length), index=times)
series = TimeSeries.from_series(pd_series).astype(np.float32)
train_series, val_series = series[:-int(length*0.8)], series[-int(length*0.2):]

# model = TFTModel(100, 100, add_relative_index=True, **tfm_kwargs)
model = TSMixerModel(100, 100, **tfm_kwargs)
model.fit(train_series, val_series=val_series, verbose=True)
model.predict(100)
```

dennisbader commented:
@daidahao 

> Another issue that might be related but perhaps best addressed in another PR. Is there a need for predict_from_dataset() to return the same dtype as the input? I imagine many users would expect the output dtype to be the same as the input, regardless of the model being used. What do you think?

Yes, that's what I was worried a bit about in the beginning as well where I mentioned that we should handle maybe handle a only subset of scenarios where the output dtype would match the input dtype. You mean both `predict()` and `predict_from_dataset()`, right?

The user can also simply cast the input series dtype to match the precision for it to be availbale. E.g. cast `series = series.astype(np.float64)` would make `precision="64-true"` available.

I'm happy to discuss further though :)

daidahao commented:
@dennisbader 

I've implemented the changes as discussed. We are producing warnings whenever a user-defined precision is provided or input time series are 16-bit. 

I've also expanded `test_mixed_precision_training()` to cover "16-mixed" and "bf16-mixed" options. I've also updated `test_auto_precision_casting()` to allow for 32-bit predictions when inputs are 16-bit and similarly `test_builtin_extended_trainer()`.

> Yes, that's what I was worried a bit about in the beginning as well where I mentioned that we should handle maybe handle a only subset of scenarios where the output dtype would match the input dtype. You mean both predict() and predict_from_dataset(), right?

Yes, I think `predict()` depends on `predict_from_dataset()`. I see it as a separate issue because the current implementation of `predict_from_dataset()` and related functions such as `residuals()` do not explicitly guarantee the return data type is the same. I notice that calling `residuals()` on `ConformalQRModel` returns `np.float64` regardless of the input data type. I could raise this as a separate issue if you think they should be the same.

daidahao commented:
@dennisbader Thanks, Dennis! I think I have addressed all your suggestions. Let me know what you think.

dennisbader reviewed (approved):
Thanks a lot for the updates @daidahao and this PR @daidahao 🚀 
All is ready now to be merged after the tests have passed 💯
